### ISSUE TITLE
Compilation fails with GCC 8.1.0

### ISSUE THREAD

**Issue by FloopCZ** [ericniebler/range-v3#826]
The corresponding outputs from the [`cmake ..`](https://github.com/ericniebler/range-v3/files/1987914/cmake.log) command and [`make`](https://github.com/ericniebler/range-v3/files/1987915/make.log) command. g++ version:

```
[floop@pear build (master)]$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/8.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --enable-multilib --disable-werror --enable-checking=release --enable-default-pie --enable-default-ssp
Thread model: posix
gcc version 8.1.0 (GCC)
```

**Comment by cpplearner** [comment]
g++ does not like

```c++
std::vector<int> v;
v = std::move(v) | push_front({1,2,3});
```

which triggers the instantiation of `std::_Bind<push_front_fn(std::_Placeholder<1>, std::initializer_list<int>)>::operator()<std::vector<int>>`. The `const volatile` overload of such `operator()` queries `std::result_of<const volatile push_front_fn&(std::vector<int>&&, const volatile std::initializer_list<int>&)>`, which eventually calls `ranges::begin(const volatile std::initializer_list<int>&)`, which delegates to `ranges::v3::_begin_::fn::impl_`, which should cause a substitution failure in [begin_end.hpp#L65](https://github.com/ericniebler/range-v3/blob/8ccd974/include/range/v3/begin_end.hpp#L65) (because `begin(std::declval<const volatile std::initializer_list<int> &>())` is invalid), but instead causes a hard error.

**Comment by cpplearner** [comment]
reported: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85765

**Comment by FloopCZ** [comment]
This is fixed in #827, thank you @CaseyCarter.
